services/miner_backend/serving/condense_app.py

```diff
@@ -0,0 +1,172 @@
+            dtype=self.dtype, device="cuda"
+        )
+        self.linear = nn.Linear(self.hidden_size * n_last_hidden_states, 4096).to(
+            dtype=self.dtype, device="cuda"
+        )
+        self.pre_condensed_tokens = nn.Parameter(
+            torch.randn(
+                1,
+                num_condense_tokens,
+                self.hidden_size,
+                dtype=self.dtype,
+                device="cuda",
+            )
+        )
+
+    def load_state_dict(self, state_dict):
+        self.pre_condensed_tokens.data = state_dict["pre_condensed_tokens"].to(
+            dtype=self.dtype, device="cuda"
+        )
+        self.linear.load_state_dict(
+            {
+                k: v.to(dtype=self.dtype, device="cuda")
+                for k, v in state_dict["linear_state_dict"].items()
+            }
+        )
+        self.norm.load_state_dict(
+            {
+                k: v.to(dtype=self.dtype, device="cuda")
+                for k, v in state_dict["norm_state_dict"].items()
+            }
+        )
+
+    @torch.no_grad()
+    def compress(self, context: str) -> torch.Tensor:
+        output = self.condense_tokenizer(
+            context,
+            return_tensors="pt",
+            add_special_tokens=False,
+            padding="max_length",
+            max_length=4096,
+            truncation=True,
+            return_attention_mask=True,
+        )
+        context_ids = output.input_ids.to(device="cuda")
+        attention_mask = output.attention_mask.to(device="cuda")
+
+        # Processing embedding and condensation
+        context_embeds = self.condense_model.get_input_embeddings()(context_ids)
+        inputs_embeds_condense = torch.cat(
+            [context_embeds, self.pre_condensed_tokens], dim=1
+        )
+        expanded_attention_mask = torch.cat(
+            [
+                attention_mask,
+                torch.ones(
+                    attention_mask.shape[0],
+                    self.num_condense_tokens,
+                    dtype=attention_mask.dtype,
+                    device=attention_mask.device,
+                ),
+            ],
+            dim=1,
+        )
+
+        # Generate condensed tokens
+        output = self.condense_model(
+            inputs_embeds=inputs_embeds_condense,
+            output_hidden_states=True,
+            attention_mask=expanded_attention_mask,
+        )
+        hidden_states = torch.cat(
+            output.hidden_states[-self.n_last_hidden_states :], dim=-1
+        )[:, -self.num_condense_tokens :, :]
+
+        return self.linear(self.norm(hidden_states)).to(torch.float32)
+
+
+def ndarray_to_base64(array: np.ndarray) -> str:
+    """Convert a NumPy array to a base64-encoded string."""
+    buffer = io.BytesIO()
+    np.save(buffer, array)
+    buffer.seek(0)
+    base64_str = base64.b64encode(buffer.read()).decode("utf-8")
+    return base64_str
+
+
+# Load state dict and get model configuration
```

Comment:
implement `def from_pretrained(self, model_id):` instead of loading by external lines